src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs

```diff
@@ -227,14 +228,86 @@ public override bool DirectoryExists(string fullPath)
 
         public override IEnumerable<string> EnumeratePaths(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
-            // Fill in implementation
-            return new string[0];
+            IEnumerable<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+
+            // Yield the paths of the StorageItems
+            foreach (IStorageItem item in storageFiles)
+            {
+                yield return item.Path;
+            }
         }
 
         public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
-            // Fill in implementation
-            return new FileSystemInfo[0];
+            IEnumerable<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+
+            // Translate the StorageFile results into FileSystemInfos
+            foreach (IStorageItem item in storageFiles)
+            {
+                if (item.IsOfType(StorageItemTypes.Folder))
+                {
+                    yield return new DirectoryInfo(item.Path);
+                }
+                else // If it is neither a File nor folder then we treat it as a File.
+                {
+                    yield return new FileInfo(item.Path);
+                }
+            }
+        }
+
+        private async static Task<IEnumerable<IStorageItem>> EnumerateFileQuery(string path, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
+        {
+            // Get a StorageFolder for "path"
+            StorageFolder folder;
+            try
+            {
+                folder = await StorageFolder.GetFolderFromPathAsync(Path.GetFullPath(path));
+            }
+            catch (FileNotFoundException e)
+            {
+                // GetFolderFromPathAsync throws FileNotFoundException but win32 throws a DirectoryNotFoundException for this case.
+                throw new DirectoryNotFoundException(e.Message, e.InnerException);
+            }
+
+            // Construct a query for the search.
+            string fileExtension = Path.GetExtension(searchPattern);
+            QueryOptions query;
+            if (fileExtension.Length > 0 && !fileExtension.Contains("?") && !fileExtension.Contains("*")) // The Query won't allow wildcards in file extensions.
+            { 
+                var fileTypes = new List<string>(1);
+                fileTypes.Add(fileExtension);
+                query = new QueryOptions(CommonFileQuery.DefaultQuery, fileTypes);
```

Comment:
Nit: Could this just use an array instead of a `List<T>` to avoid the extra object allocation? e.g.

```csharp
var fileTypes = new string[] { fileExtension };
query = new QueryOptions(CommonFileQuery.DefaultQuery, fileTypes);
```